only do a full bootstrap on startup or the migration process

calellowitz opened the pull request:
## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
The bootstrap that happens every three hours causes a pretty notable memory memory spike on pillow machines. This is in part because it is a memory heavy process that will only get more so as more people use UCRs, but in part because all pillow processes on a machine basically do it at the same time, so it ends up using about 20X memory. https://app.datadoghq.com/dashboard/ewu-jyr-udt/change-feeds-pillows?fullscreen_end_ts=1654803637572&fullscreen_paused=false&fullscreen_section=overview&fullscreen_start_ts=1654630837572&fullscreen_widget=265293246&from_ts=1654630824488&to_ts=1654803624488&live=true

This changes the pillow to only do a full bootstrap in processes that run migrations. Others just use the incremental update to data sources that was occurring anyway. I added the config validation from the full bootstrap to the incremental process to make the two more equivalent. 

The one place it might represent a real change is that the regular bootstrap does remove configs from domains set as `migration_in_progress`. This only happens once every 3 hours (during the bootstrap) so there is still a decent lag after that occurs, and the migration flags should prevent any case changes, meaning no cases from those domains should hit the pillow to begin with, but that is a change.

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
This should not change any existing behavior, and to the extent the included configs differ, it is excluding invalid configurations, which should reduce errors.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

calellowitz committed:
only do a full bootstrap on startup or the migration process

snopoke reviewed (commented):
One potential issue I see here is that this won't remove datasources, it will only add new ones in. For example if a DS is edited and the new config is invalid the old config will continue to be used (maybe that's fine but there could be other scenarios).

I think you should get the difference between the 'new_data_sources` and `filtered_data_sources` and make sure that any that were filtered out are also removed from the already configured adapters.

calellowitz committed:
fix typo

calellowitz committed:
remove invalid configs

calellowitz committed:
update calls in test

calellowitz committed:
use proper syntax for set comprehension

calellowitz commented:
> One potential issue I see here is that this won't remove datasources, it will only add new ones in. For example if a DS is edited and the new config is invalid the old config will continue to be used (maybe that's fine but there could be other scenarios).
> 
> I think you should get the difference between the 'new_data_sources`and`filtered_data_sources` and make sure that any that were filtered out are also removed from the already configured adapters.

Great point. I added that as well

snopoke commented on corehq/apps/userreports/pillow.py:
```diff
@@ -273,9 +276,11 @@ def _update_modified_since(self, timestamp):
-        self._add_data_sources_to_table_adapters(new_data_sources)
+        filtered_data_sources = self.get_filtered_configs(new_data_sources)
+        invalid_data_sources = {ds._id for ds in new_data_sources} - {ds._id for ds in filtered_data_sources}
+        self._add_data_sources_to_table_adapters(filtered_data_sources)
```
```suggestion
        self._add_data_sources_to_table_adapters(filtered_data_sources, invalid_data_sources)
```

calellowitz committed:
pass new argument

calellowitz committed:
respect empty list

dannyroberts commented:
@calellowitz is your read that the Stickler "could not clone repository" thing is just a temporary failure on Sitckler's end? Is it worth attempting to rerun that somehow to make sure its previous suggestions have been addressed? (Fine with "not really" if that's the answer.)

calellowitz commented:
> @calellowitz is your read that the Stickler "could not clone repository" thing is just a temporary failure on Sitckler's end? Is it worth attempting to rerun that somehow to make sure its previous suggestions have been addressed? (Fine with "not really" if that's the answer.)

Not sure what previous suggestions you were referring to, but I kicked off the stickler bot again (if I knew I could just press a button to do so, I would have done that a long time ago), and it looks like it passed.
